src/libcore/yk_swt_impl.c

```diff
@@ -29,52 +29,52 @@ static __thread size_t trace_buf_len = 0;
 // The allocation capacity of the trace buffer (in elements).
 static __thread size_t trace_buf_cap = 0;
 // Is the current thread tracing?
-static __thread bool tracing = false;
-// This flag is set when the contents of the trace buffer may not be valid.
-static __thread volatile atomic_bool trace_invalid = false;
+// true = we are tracing, false = we are not tracing or an error occurred.
+static __thread volatile atomic_bool tracing = false;
 
 // Start tracing on the current thread.
-// A new trace buffer is allocated and MIR locations are written into it. The
-// current thread must not already be tracing.
+// A new trace buffer is allocated and MIR locations will be written into it on
+// subsequent calls to `yk_swt_rec_loc_impl`. The current thread must not
+// already be tracing.
 void
 yk_swt_start_tracing_impl() {
-    if (tracing) {
-        errx(EXIT_FAILURE, "%s: thread already tracing", __func__);
-    }
-
     trace_buf = calloc(TL_TRACE_INIT_CAP, sizeof(struct mir_loc));
     if (trace_buf == NULL) {
         err(EXIT_FAILURE, "%s: calloc: ", __func__);
     }
 
     trace_buf_cap = TL_TRACE_INIT_CAP;
-    tracing = true;
-    atomic_store_explicit(&trace_invalid, false, memory_order_relaxed);
+    atomic_store_explicit(&tracing, true, memory_order_relaxed);
 }
 
 // Record a location into the trace buffer if tracing is enabled on the current thread.
 void
 yk_swt_rec_loc_impl(uint64_t crate_hash, uint32_t def_idx, uint32_t bb_idx)
 {
-    if (!tracing) {
+    if (!atomic_load_explicit(&tracing, memory_order_relaxed)) {
         return;
     }
 
     // Check if we need more space and reallocate if necessary.
     if (trace_buf_len == trace_buf_cap) {
         if (trace_buf_cap >= SIZE_MAX - TL_TRACE_REALLOC_CAP) {
-            errx(EXIT_FAILURE, "%s: trace capacity would overflow", __func__);
+            // Trace capacity would overflow.
+            atomic_store_explicit(&tracing, false, memory_order_relaxed);
+            return;
         }
         size_t new_cap = trace_buf_cap + TL_TRACE_REALLOC_CAP;
 
         if (new_cap > SIZE_MAX / sizeof(struct mir_loc)) {
-            errx(EXIT_FAILURE, "%s: new buffer size would overflow", __func__);
+            // New buffer size would overflow.
```

Comment:
Should/could we free the buffer at this point? Similarly for the `realloc` failure below.

Reply:
It will be freed when tracing stops. Until then the user is oblivious to the error anyway.

Reply:
Yes, but do we want to waste memory in this way? Especially if realloc fails -- the user is in trouble, and we might be able to help them out a little.

Reply:
Unless I'm mistaken, it's not safe (at least not without a mutex or similar). The stack overflow handler could cause us to re-enter here in the event that SIGSEGV and SIGBUS are delivered in quick succession.

freeing and NULLing would need to be performed atomically.

This is why I free and NULL when we stop the tracer. The stack overflow handler can't end up in there.

Reply:
> do we want to waste memory in this way

I don't understand why memory would be wasted though? The same memory is used whether we free here or later.

Reply:
> Unless I'm mistaken, it's not safe (at least not without a mutex or similar).

Good point.